rasa/nlu/data_augmentation/augmenter.py

```diff
@@ -0,0 +1,699 @@
+
+    The tokenizer is required for the augmentation strategy based on maximum vocabulary expansion where the paraphrases
+    as well as the originally supplied training data need to be tokenized.
+
+    Args:
+         config_path: Path to the config file.
+
+    Returns:
+        The tokenizer from the config file or the WhitespaceTokenizer if no tokenizer was found in the config file.
+    """
+    config = rasa.shared.utils.io.read_config_file(config_path)
+    pipeline = config.get("pipeline", [])
+    tokenizer_config = {"name": "WhitespaceTokenizer"}
+    for component in pipeline:
+        if component.get("name", "").lower().endswith("tokenizer"):
+            tokenizer_config = component
+            break
+
+    tokenizer = ComponentBuilder().create_component(
+        tokenizer_config, RasaNLUModelConfig(config)
+    )
+
+    return tokenizer
+
+
+def _create_paraphrase_pool(
+    paraphrases: TrainingData,
+    intents_to_augment: Set[Text],
+    min_paraphrase_sim_score: float,
+    max_paraphrase_sim_score: float,
+) -> Dict[Text, List]:
+    """Determines all suitable paraphrases for data augmentation for the given intents.
+
+    Args:
+        paraphrases: The paraphrases for data augmentation.
+        intents_to_augment: The intents for which to perform data augmentation.
+        min_paraphrase_sim_score: Accept/Reject minimum similarity threshold for individual paraphrases.
+        max_paraphrase_sim_score: Accept/Reject maximum similarity threshold for individual paraphrases.
+
+    Returns:
+        The pool of suitable paraphrases for data augmentation.
+    """
+
+    paraphrase_pool = collections.defaultdict(list)
+    for paraphrase_msg in paraphrases.intent_examples:
+
+        paraphrases_for_example = (
+            paraphrase_msg.get(METADATA, {})
+            .get(METADATA_EXAMPLE, {})
+            .get("paraphrases", {})
+        )
+        if (
+            paraphrase_msg.get(INTENT) not in intents_to_augment
+            or not paraphrases_for_example
+        ):
+            continue
+
+        paraphrase_scores = (
+            paraphrase_msg.get(METADATA, {}).get(METADATA_EXAMPLE, {}).get("scores", {})
+        )
+
+        if not paraphrase_scores:
+            logger.debug(
+                f"""Skipping "{paraphrase_msg.get(TEXT)}" as no similarity scores were found for its paraphrases."""
+            )
+
+        for paraphrase, score in zip(paraphrases_for_example, paraphrase_scores):
+            if (
+                paraphrase == ""
+                or float(score) < min_paraphrase_sim_score
+                or float(score) > max_paraphrase_sim_score
+            ):
+                continue
+
+            # Create Message-compatible data representation for the paraphrases
+            data = {TEXT: paraphrase, INTENT: paraphrase_msg.get(INTENT)}
+            paraphrase_pool[paraphrase_msg.get(INTENT)].append(Message(data=data))
```

Comment:
At this stage we're only collecting all paraphrases from the supplied file (per `intent_to_augment`) that pass the basic min/max score thresholds. Choosing paraphrases based on `max_vocab_expansion` and random sampling (with the subsequent creation of a `TrainingData` obj) happens at a later stage.

Its easier to access specific examples for a given intent with a dictionary, otherwise I'd always to either loop or apply a filter to the training data (which in turn does the loop), so I thought the use of a dictionary more explicit and easier to use.

wdyt?

Reply:
Okay 👍  Let's appropriately have tests to make  sure that the keys of this dict have all intents, test other related things to this structure.